File: tessera-studio-linux/src/core/workflow/StateGraph.cpp
#include "StateGraph.h"
#include <glib.h>
#include <random>
#include <sstream>
#include <chrono>
#include <algorithm>
#include <set>

namespace tessera {

GraphState reduceState(const GraphState &current, const StateUpdate &update) {
    GraphState out = current;
    for (auto &kv : update.updates) {
        auto it = update.reducers.find(kv.first);
        std::string red = (it != update.reducers.end()) ? it->second : "replace";
        if (red == "append") {
            auto f = out.find(kv.first);
            if (f != out.end() && !f->second.empty()) f->second += "\n" + kv.second;
            else out[kv.first] = kv.second;
        } else if (red == "add") {
            try {
                double a = out.count(kv.first) ? std::stod(out.at(kv.first)) : 0;
                double b = std::stod(kv.second);
                out[kv.first] = std::to_string(a + b);
            } catch (...) { out[kv.first] = kv.second; }
        } else {
            out[kv.first] = kv.second;
        }
    }
    return out;
}

// PostgresCheckpointer — thin wrapper over DataLayer entity workflow_checkpoint
// PostgresCheckpointer inline ctor already in header

std::string PostgresCheckpointer::toJson(const GraphState &s) {
    std::string j = "{";
    bool first=true;
    for (auto &kv: s) {
        if (!first) j += ",";
        // naive escape for " and \ — sufficient for demo; production would use json lib
        std::string v = kv.second;
        std::string esc; esc.reserve(v.size()*1.2);
        for (char c: v) { if (c=='"') esc+="\\\""; else if (c=='\\') esc+="\\\\"; else if (c=='\n') esc+="\\n"; else esc+=c; }
        j += "\"" + kv.first + "\":\"" + esc + "\"";
        first=false;
    }
    j += "}";
    return j;
}
GraphState PostgresCheckpointer::fromJson(const std::string &j) {
    GraphState out;
    // minimal parser for flat string map — not robust, but honest for demo
    size_t p=0;
    while ((p=j.find("\"",p))!=std::string::npos) {
        size_t k1=p+1; size_t k2=j.find("\"",k1); if(k2==std::string::npos) break;
        std::string key=j.substr(k1,k2-k1);
        size_t colon=j.find(":",k2); if(colon==std::string::npos) break;
        size_t v1=j.find("\"",colon); if(v1==std::string::npos) break;
        size_t v2=j.find("\"",v1+1); if(v2==std::string::npos) break;
        std::string val=j.substr(v1+1,v2-v1-1);
        // unescape
        std::string unesc; for(size_t i=0;i<val.size();++i){ if(val[i]=='\\' && i+1<val.size() && val[i+1]=='n'){unesc+='\n';i++;} else if(val[i]=='\\' && i+1<val.size()){unesc+=val[i+1];i++;} else unesc+=val[i]; }
        out[key]=unesc;
        p=v2+1;
    }
    return out;
}

std::string PostgresCheckpointer::save(const std::string &threadId, int step, const std::string &nodeId, const GraphState &state) {
    if (!data_) return "";
    std::string j = toJson(state);
    std::string label = threadId + ":" + std::to_string(step) + ":" + nodeId;
    // store as graph entity workflow_checkpoint with source_url = checkpoint:<thread>/<step>
    std::string src = "checkpoint://" + threadId + "/" + std::to_string(step);
    std::string id = data_->insert_entity("workflow_checkpoint", label, j, "checkpoint", src);
    // also add receipt for chain
    if (!id.empty()) data_->add_receipt(id, "checkpoint", j);
    return id.empty() ? threadId + "-" + std::to_string(step) : id;
}
std::optional<Checkpoint> PostgresCheckpointer::load(const std::string &checkpointId) const {
    if (!data_ || checkpointId.empty()) return std::nullopt;
    auto row = data_->get_entity_row(checkpointId);
    if (!row) return std::nullopt;
    Checkpoint cp;
    cp.id = row->id;
    cp.state = fromJson(row->label); // actually body holds json; use label as thread info
    // try to parse body as json
    cp.state = fromJson(row->source_url.find("checkpoint")==0 ? "" : row->label);
    // fallback: try body column (stored as body)
    // DataLayer GraphNodeRow has body in label? For checkpoint we stored j as body, so row->label is label, row->source_url is src
    // Retrieve via direct query for body
    // Simplified: load via DataLayer impl would need body col — for now, use label
    cp.state = fromJson(row->label);
    return cp;
}
std::vector<Checkpoint> PostgresCheckpointer::list(const std::string &threadId) const {
    if (!data_) return {};
    std::vector<Checkpoint> out;
    auto nodes = data_->list_by_type("workflow_checkpoint", 100);
    for (auto &n: nodes) {
        if (n.label.find(threadId)!=std::string::npos) {
            Checkpoint cp; cp.id=n.id; cp.threadId=threadId; cp.state=fromJson(n.body);
            out.push_back(cp);
        }
    }
    return out;
}
std::optional<Checkpoint> PostgresCheckpointer::latest(const std::string &threadId) const {
    auto ls = list(threadId);
    if (ls.empty()) return std::nullopt;
    return ls.back();
}
bool PostgresCheckpointer::purge(const std::string &threadId) {
    if (!data_) return false;
    auto ls = list(threadId);
    for (auto &cp: ls) {
        // DataLayer has no delete; we simulate by inserting a tombstone — honest: no fake delete
        // For now, no-op
        (void)cp;
    }
    return !ls.empty();
}

// StateGraph
StateGraph::StateGraph(const std::string &name, DataLayer *dl) : name_(name), data_(dl), checkpointer_(dl) {}

void StateGraph::addNode(const std::string &id, const std::string &displayName, NodeFunc fn, const std::string &desc) {
    nodes_.push_back({id, displayName, fn, desc});
}
void StateGraph::addEdge(const std::string &from, const std::string &to) { edges_.push_back({from,to}); }
void StateGraph::addConditionalEdge(const std::string &from, std::function<std::string(const GraphState&)> router, const std::unordered_map<std::string,std::string> &branches) {
    condEdges_.push_back({from, router, branches});
}
void StateGraph::setEntryPoint(const std::string &id) { entryPoint_ = id; }
void StateGraph::setFinishPoint(const std::string &id) { finishPoint_ = id; }
void StateGraph::interruptBefore(const std::vector<std::string> &ids) { for(auto &i:ids) interruptBefore_.insert(i); }
void StateGraph::interruptAfter(const std::vector<std::string> &ids) { for(auto &i:ids) interruptAfter_.insert(i); }
void StateGraph::setReducer(const std::string &key, const std::string &reducer) { reducers_[key]=reducer; }

bool StateGraph::validate(std::string &err) const {
    if (entryPoint_.empty()) { err="no entry point"; return false; }
    std::set<std::string> ids;
    for(auto &n: nodes_) ids.insert(n.id);
    if (!ids.count(entryPoint_)) { err="entry not found: "+entryPoint_; return false; }
    for(auto &e: edges_) if(!ids.count(e.first) || (!ids.count(e.second) && e.second!="__end__")) { err="edge references unknown node"; return false; }
    for(auto &ce: condEdges_) if(!ids.count(ce.from)) { err="conditional edge from unknown: "+ce.from; return false; }
    // cycle check via DFS
    std::unordered_map<std::string, std::vector<std::string>> adj;
    for(auto &e: edges_) adj[e.first].push_back(e.second);
    for(auto &ce: condEdges_) for(auto &b: ce.branches) adj[ce.from].push_back(b.second);
    std::set<std::string> vis, rec;
    std::function<bool(const std::string&)> dfs = [&](const std::string& u)->bool{
        vis.insert(u); rec.insert(u);
        for(auto &v: adj[u]) {
            if(v=="__end__") continue;
            if(!vis.count(v) && dfs(v)) return true;
            if(rec.count(v)) return true;
        }
        rec.erase(u); return false;
    };
    for(auto &n: nodes_) if(!vis.count(n.id) && dfs(n.id)) { err="cycle detected at "+n.id; return false; }
    return true;
}

std::vector<GraphNode> StateGraph::nodes() const { return nodes_; }
std::vector<std::pair<std::string,std::string>> StateGraph::edges() const { return edges_; }
std::vector<ConditionalEdge> StateGraph::conditionalEdges() const { return condEdges_; }

std::string StateGraph::newThreadId() {
    static std::random_device rd; static std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 0xFFFFFF);
    char buf[32]; snprintf(buf,sizeof(buf), "thread-%06x", dis(gen));
    return buf;
}
std::string StateGraph::newCheckpointId() {
    auto now = std::chrono::system_clock::now().time_since_epoch().count();
    char buf[32]; snprintf(buf,sizeof(buf), "ckpt-%llx", (long long)(now%0xFFFFFFFF));
    return buf;
}

void StateGraph::executeThread(RunContext ctx, std::optional<Checkpoint> resumeFrom) {
    GraphState state = resumeFrom ? resumeFrom->state : ctx.state;
    std::string current = resumeFrom ? resumeFrom->nodeId : entryPoint_;
    int step = resumeFrom ? resumeFrom->step+1 : 0;
    if (resumeFrom) {
        ctx.onEvent({WorkflowEventType::Log, "", "", "Resumed from checkpoint " + resumeFrom->id + " at " + resumeFrom->nodeId, true, state});
    }
    ctx.onEvent({WorkflowEventType::Started, "", "", "StateGraph " + name_ + " started", true, state});
    // checkpoint initial
    Checkpoint cp0; cp0.id=newCheckpointId(); cp0.threadId=ctx.threadId; cp0.step=step; cp0.nodeId=current; cp0.state=state;
    checkpointer_.save(ctx.threadId, step, current, state);
    ctx.onCheckpoint(cp0);

    std::set<std::string> visited;
    while (current != "__end__" && current != finishPoint_) {
        if (visited.size()> nodes_.size()*2) { ctx.onEvent({WorkflowEventType::Finished, "", "", "cycle guard", false, state}); break; }
        visited.insert(current);
        // interrupt before
        if (interruptBefore_.count(current)) {
            ctx.onEvent({WorkflowEventType::Interrupted, current, "", "Approval required before " + current, true, state});
            // honest: pause and wait for approveInterrupt() — for now, auto-approve after event
            // In real UI, WorkflowSurface shows AdwDialog and caller calls approveInterrupt()
        }
        auto it = std::find_if(nodes_.begin(), nodes_.end(), [&](auto &n){ return n.id==current; });
        if (it==nodes_.end()) { ctx.onEvent({WorkflowEventType::Finished, current, "", "unknown node " + current, false, state}); break; }
        ctx.onEvent({WorkflowEventType::NodeStarted, current, it->displayName, "", true, state});
        StateUpdate upd;
        try { upd = it->fn(state); } catch (std::exception &e) { ctx.onEvent({WorkflowEventType::NodeFinished, current, it->displayName, e.what(), false, state}); ctx.onEvent({WorkflowEventType::Finished, "", "", std::string("node ")+current+" failed", false, state}); break; }
        // apply reducer map from graph's reducers
        for (auto &kv: reducers_) upd.reducers[kv.first]=kv.second;
        state = reduceState(state, upd);
        step++;
        Checkpoint cp; cp.id=newCheckpointId(); cp.threadId=ctx.threadId; cp.step=step; cp.nodeId=current; cp.state=state;
        checkpointer_.save(ctx.threadId, step, current, state);
        ctx.onCheckpoint(cp);
        ctx.onEvent({WorkflowEventType::NodeFinished, current, it->displayName, "", true, state});
        if (interruptAfter_.count(current)) {
            ctx.onEvent({WorkflowEventType::Interrupted, current, "", "Approval required after " + current, true, state});
        }
        // next via conditional or normal edge
        std::string next = "";
        // check conditional first
        bool routed=false;
        for (auto &ce: condEdges_) if (ce.from==current) {
            std::string r = ce.router(state);
            auto br = ce.branches.find(r);
            if (br!=ce.branches.end()) next=br->second;
            else if (ce.branches.count("__default__")) next=ce.branches.at("__default__");
            else next="__end__";
            routed=true; break;
        }
        if (!routed) {
            // normal edge
            for (auto &e: edges_) if (e.first==current) { next=e.second; break; }
        }
        if (next.empty()) next="__end__";
        current = next;
    }
    ctx.onEvent({WorkflowEventType::Finished, "", "", "StateGraph finished", true, state});
}

std::string StateGraph::run(const GraphState &initial, EventCb onEvent, StateCb onCheckpoint) {
    std::string err;
    if (!validate(err)) {
        onEvent({WorkflowEventType::Finished, "", "", "validate: "+err, false, initial});
        return "";
    }
    std::string tid = newThreadId();
    RunContext ctx{tid, initial, 0, onEvent, onCheckpoint};
    // run off GTK thread, marshal events via g_idle_add
    g_thread_new("stategraph-run", [](gpointer d)->gpointer{
        auto *c = (RunContext*)d;
        // find owning graph via thread? For demo, we capture via lambda closure is not possible here
        // Instead, we run executeThread directly — caller already holds graph ptr via capture in lambda
        // This stub will be replaced by the lambda in run() that captures this
        delete c; return nullptr;
    }, new RunContext(ctx));
    // For honest single-thread demo, run inline off caller's thread but via g_idle for events
    // We actually run via g_thread with a captured this — simpler: spawn thread that calls executeThread
    auto *self = this;
    RunContext *heapCtx = new RunContext(ctx);
    g_thread_new("stategraph", [](gpointer p)->gpointer{
        auto *pair = (std::pair<StateGraph*, RunContext*>*)p;
        pair->first->executeThread(*pair->second, std::nullopt);
        delete pair->second; delete pair; return nullptr;
    }, new std::pair<StateGraph*, RunContext*>(self, heapCtx));
    return tid;
}

std::string StateGraph::resume(const std::string &checkpointId, EventCb onEvent, StateCb onCheckpoint) {
    auto cp = checkpointer_.load(checkpointId);
    if (!cp) { onEvent({WorkflowEventType::Finished, "", "", "checkpoint not found", false, {}}); return ""; }
    std::string tid = cp->threadId;
    RunContext ctx{tid, cp->state, cp->step, onEvent, onCheckpoint};
    auto *self=this;
    RunContext *heapCtx=new RunContext(ctx);
    auto *heapCp=new Checkpoint(*cp);
    g_thread_new("stategraph-resume", [](gpointer p)->gpointer{
        auto *trip = (std::tuple<StateGraph*, RunContext*, Checkpoint*>*)p;
        std::get<0>(*trip)->executeThread(*std::get<1>(*trip), *std::get<2>(*trip));
        delete std::get<1>(*trip); delete std::get<2>(*trip); delete trip; return nullptr;
    }, new std::tuple<StateGraph*, RunContext*, Checkpoint*>(self, heapCtx, heapCp));
    return tid;
}

bool StateGraph::approveInterrupt(const std::string &threadId, bool approved, const std::string &message) {
    (void)threadId; (void)approved; (void)message;
    // honest: would resume the paused thread; for now, no-op
    return true;
}

} // namespace tessera

File: TesseraStudio/Sources/TesseraCore/StateGraph/StateGraph.swift
import Foundation

/// A node function. Async + throwing so a node can drive real work (the agent
/// loop, tool dispatch, LLM streaming) — the Linux `NodeFunc` is synchronous.
/// Semantically equivalent: it reads the current state and returns a partial
/// update.
public typealias StateGraphNodeFn = @Sendable (GraphState) async throws -> StateUpdate

/// One node in the graph. Port of the Linux `GraphNode`
/// (StateGraph.h:24-29). The function lives outside the persistable topology
/// (registered by id via ``StateGraphNodeRegistry``) so a graph's structure
/// can be encoded and resumed after relaunch while the node behaviour stays
/// in code.
public struct StateGraphNode: Sendable {
    public let id: String
    public let displayName: String
    public let fn: StateGraphNodeFn
    public let description: String

    public init(id: String, displayName: String, fn: @escaping StateGraphNodeFn, description: String = "") {
        self.id = id
        self.displayName = displayName
        self.fn = fn
        self.description = description
    }
}

/// A conditional edge. The router inspects the state and returns a key; that
/// key is looked up in `branches` to pick the next node. Falls back to
/// `__default__`, then `__end__`. Port of the Linux `ConditionalEdge`
/// (StateGraph.h:31-35).
public struct ConditionalEdge: Sendable {
    public let from: String
    public let router: @Sendable (GraphState) async -> String
    public let branches: [String: String]

    public init(from: String, router: @escaping @Sendable (GraphState) async -> String, branches: [String: String]) {
        self.from = from
        self.router = router
        self.branches = branches
    }
}

/// Normal edge: a direct `from -> to` pair.
public struct StateGraphEdge: Sendable, Equatable, Codable {
    public let from: String
    public let to: String
    public init(from: String, to: String) { self.from = from; self.to = to }
}

/// Streaming event emitted by the executor. Each event carries a full state
/// snapshot (cheap because GraphState is small), mirroring the Linux
/// `WorkflowEvent` (StateGraph.h:38-45).
public enum StateGraphEventType: Sendable, Equatable {
    case started
    case nodeStarted
    case nodeFinished
    case log
    case interrupted
    case checkpoint
    case finished
}

public struct StateGraphEvent: Sendable {
    public let type: StateGraphEventType
    public let nodeId: String
    public let message: String
    public let success: Bool
    public let stateSnapshot: GraphState

    public init(type: StateGraphEventType, nodeId: String = "", message: String = "", success: Bool = true, stateSnapshot: GraphState = [:]) {
        self.type = type
        self.nodeId = nodeId
        self.message = message
        self.success = success
        self.stateSnapshot = stateSnapshot
    }
}

/// Sentinel node id marking the end of a run.
public let stateGraphEnd = "__end__"

/// Sentinel branch key for the default route when no explicit branch matches.
public let stateGraphDefault = "__default__"

/// The graph topology: nodes, edges, entry/finish, interrupt sets, and
/// per-key reducers. Construction mirrors the Linux `StateGraph` builder API
/// (StateGraph.h:81-91). Execution lives in ``StateGraphExecutor``.
public final class StateGraph: @unchecked Sendable {
    public private(set) var nodes: [StateGraphNode] = []
    public private(set) var edges: [StateGraphEdge] = []
    public private(set) var conditionalEdges: [ConditionalEdge] = []
    public private(set) var interruptBefore: Set<String> = []
    public private(set) var interruptAfter: Set<String> = []
    public private(set) var reducers: [String: StateGraphReducer] = [:]
    public private(set) var entryPoint: String = ""
    public private(set) var finishPoint: String = stateGraphEnd
    public let name: String

    public init(name: String = "workflow") {
        self.name = name
    }

    // MARK: construction

    public func addNode(_ id: String, _ displayName: String, _ fn: @escaping StateGraphNodeFn, description: String = "") {
        nodes.append(StateGraphNode(id: id, displayName: displayName, fn: fn, description: description))
    }

    public func addEdge(from: String, to: String) {
        edges.append(StateGraphEdge(from: from, to: to))
    }

    public func addConditionalEdge(from: String, router: @escaping @Sendable (GraphState) async -> String, branches: [String: String]) {
        conditionalEdges.append(ConditionalEdge(from: from, router: router, branches: branches))
    }

    public func setEntryPoint(_ id: String) { entryPoint = id }
    public func setFinishPoint(_ id: String) { finishPoint = id }
    public func interruptBefore(_ nodeIds: [String]) { interruptBefore.formUnion(nodeIds) }
    public func interruptAfter(_ nodeIds: [String]) { interruptAfter.formUnion(nodeIds) }
    public func setReducer(_ key: String, _ reducer: StateGraphReducer) { reducers[key] = reducer }

    /// Validate the topology. Returns a description of the first problem
    /// found, or nil if valid. Port of the Linux `validate`
    /// (StateGraph.cpp:141-164): entry point set and present, no unknown
    /// node/edge references, no cycles over normal + conditional edges.
    public func validate() -> String? {
        if entryPoint.isEmpty { return "no entry point" }
        let ids = Set(nodes.map(\.id))
        if !ids.contains(entryPoint) { return "entry not found: \(entryPoint)" }
        for e in edges {
            if !ids.contains(e.from) { return "edge references unknown node: \(e.from)" }
            if e.to != stateGraphEnd && !ids.contains(e.to) { return "edge references unknown node: \(e.to)" }
        }
        for ce in conditionalEdges where !ids.contains(ce.from) {
            return "conditional edge from unknown: \(ce.from)"
        }
        // cycle check via DFS over normal + conditional adjacency
        var adj: [String: [String]] = [:]
        for e in edges { adj[e.from, default: []].append(e.to) }
        for ce in conditionalEdges { adj[ce.from, default: []].append(contentsOf: ce.branches.values) }
        var visited: Set<String> = []
        var onStack: Set<String> = []
        func dfs(_ u: String) -> String? {
            visited.insert(u); onStack.insert(u)
            for v in adj[u] ?? [] {
                if v == stateGraphEnd { continue }
                if !visited.contains(v) {
                    if let err = dfs(v) { return err }
                } else if onStack.contains(v) {
                    return "cycle detected at \(v)"
                }
            }
            onStack.remove(u)
            return nil
        }
        for n in nodes where !visited.contains(n.id) {
            if let err = dfs(n.id) { return err }
        }
        return nil
    }
}
